src/Library/WebServer.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace Microsoft.IstioMixerPlugin.Library
+{
+    using Microsoft.IstioMixerPlugin.Common;
+    using Microsoft.IstioMixerPlugin.Library.Inputs;
+    using System;
+    using System.Net;
+    using System.Runtime.Serialization.Json;
+
+    public class WebServer
+    {
+        private HttpListener listener;
+        private readonly Configuration config;
+        private readonly Object syncObject = new Object();
+        internal bool IsRunning
+        {
+            get
+            {
+                return this.listener != null ? this.listener.IsListening : false;
+            }
+        }
+
+        public WebServer(string config)
+        {
+            if (!HttpListener.IsSupported)
+            {
+                string logLine = FormattableString.Invariant($"HttpListener is not supported");
+                Diagnostics.LogError(logLine);
+                throw new InvalidOperationException(logLine);
+            }
+
+            if (String.IsNullOrEmpty(config))
+            {
+                string logLine = FormattableString.Invariant($"config missing");
+                Diagnostics.LogError(logLine);
+                throw new ArgumentException(logLine);
+            }
+
+            this.config = new Configuration(config);
+        }
+
+        public void Start()
+        {
+            if (this.IsRunning)
+            {
+                Diagnostics.LogInfo("WebServer already running .. exiting");
+                return;
+            }
+            // Create a listener.
+            this.listener = new HttpListener();
+            this.listener.Prefixes.Add(this.config.HttpListenerPrefix);
+
+            try
+            {
+                this.listener.Start();
+                listener.BeginGetContext(new AsyncCallback(this.ListenerCallbackAsync), this.listener);
+            }
+            catch (Exception e)
+            {
+                Diagnostics.LogError(e.ToString());
+                throw e;
+            }
+
+            Diagnostics.LogInfo(FormattableString.Invariant($"Webserver running"));
+        }
+
+        public void Stop()
+        {
+            if (!this.IsRunning)
+            {
+                Diagnostics.LogInfo("WebServer already stopped .. exiting");
+                return;
+            }
+
+            try
+            {
+                this.listener.Close();
+            }
+            catch (Exception e)
+            {
+                Diagnostics.LogError(e.ToString());
+                throw e;
+            }
+            Diagnostics.LogInfo("WebServer stopped from taking any new requests");
+        }
+
+        internal void ListenerCallbackAsync(IAsyncResult result)
+        {
+            if (this.IsRunning)
```

Comment:
Should we not call EndGetContext one last time if we're in this method but IsRunning == false? I wonder if some internal context might not be cleaned up.

Reply:
if we endup here after the object has been closed through the Stop method then it will be an interesting situation. Since the object has been disposed already we cannot call EndGetState, as we will get and exception : 

Unhandled Exception: System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.HttpListener'.
   at System.Net.HttpListener.CheckDisposed()
   at System.Net.HttpListener.EndGetContext(IAsyncResult asyncResult)

I hope that the correct cleanup is being done inside the listener, since calling Close() causes the processing of any request to stop, and resources to be disposed, we are unable to control what happens to inflight requests.